plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -108,55 +111,62 @@ static AWSCredentialsProvider buildCredentials(Logger logger, Ec2ClientSettings
-        if ((clientReference != null) && clientReference.tryIncRef()) {
-            return clientReference;
-        }
-        synchronized (this) {
-            if ((clientReference != null) && clientReference.tryIncRef()) {
-                return clientReference;
-            }
-            if (clientSettings == null) {
-                throw new IllegalArgumentException("Missing ec2 client configs.");
-            }
-            final AmazonEc2Reference clientReference = new AmazonEc2Reference(buildClient(clientSettings));
-            clientReference.incRef();
-            this.clientReference = clientReference;
-            return clientReference;
+        final LazyAmazonEc2Reference clientReference = this.lazyClientReference.get();
+        if (clientReference == null) {
+            throw new IllegalStateException("Missing ec2 client configs");
         }
+        return clientReference.getOrCompute();
     }
 
-
     /**
-     * Refreshes the settings for the AmazonEC2 client. New clients will be build
-     * using these new settings. Old client is usable until released. On release it
+     * Refreshes the settings for the AmazonEC2 client. The new client will be build
+     * using these new settings. The old client is usable until released. On release it
      * will be destroyed instead of being returned to the cache.
      */
     @Override
-    public synchronized Ec2ClientSettings refreshAndClearCache(Ec2ClientSettings clientSettings) {
-        // shutdown all unused clients
-        // others will shutdown on their respective release
-        releaseCachedClient();
-        final Ec2ClientSettings prevSettings = this.clientSettings;
-        this.clientSettings = clientSettings;
-        return prevSettings;
+    public void refreshAndClearCache(Ec2ClientSettings clientSettings) {
+        final LazyAmazonEc2Reference newClient = new LazyAmazonEc2Reference(() -> buildClient(clientSettings));
+        final LazyAmazonEc2Reference oldClient = this.lazyClientReference.getAndSet(newClient);
+        if (oldClient != null) {
+            oldClient.clear();
+        }
     }
 
     @Override
     public void close() {
-        releaseCachedClient();
-    }
-
-    private synchronized void releaseCachedClient() {
-        if (this.clientReference == null) {
-            return;
+        final LazyAmazonEc2Reference clientReference = this.lazyClientReference.getAndSet(null);
+        if (clientReference != null) {
+            clientReference.clear();
         }
-        // the client will shutdown when it will not be used anymore
-        this.clientReference.decRef();
-        // clear the cached client, it will be build lazily
-        this.clientReference = null;
         // shutdown IdleConnectionReaper background thread
         // it will be restarted on new client usage
         IdleConnectionReaper.shutdown();
     }
 
+    private static class LazyAmazonEc2Reference extends Lazy<AmazonEc2Reference, Exception> {
```

Comment:
rather than a class for this, can we add a Consumer for the value on get/compute to increment the ref count? Then lazy can be final

Reply:
done, I think it's cleaner like you proposed since you don't have to study the super implementation when extending.